### ISSUE TITLE
Unit testing

### ISSUE THREAD

**Issue by acfogarty** [espressopp/espressopp#92]
We have some unit testing in espresso++ but it looks like no one except @MrTheodor is really adding new tests. Recently I was trying to write some tests for a big bunch of new code I want to push. Just wanted to check what should we think about when writing a unit test in espresso++?

I was thinking:
- test for backward compatibility when someone else pushes new code 
- test should be fast
- not have big input or output files
- cover as much of the class as possible, including different conditional pathways

Anything else?

**Comment by niktre** [comment]
maybe a straightforward thing, but I would say that the test should be reproducible (though it is not always possible)

**Comment by acfogarty** [comment]
@niktre What do you mean by reproducible, exactly? If a unit test isn't reproducible, then it will always fail, right? And that will make it useless  as a unit test.

**Comment by acfogarty** [comment]
Some of the existing unit tests are pretty slow (pi_water and polymer_melt_tabulated are 30-40 seconds each). Moreover they don't seem to actually test for anything except "runs for N steps without crashing". Does anyone know if there's some reason they're written that way that I'm not understanding? If not I'll just reduce the number of steps.

**Comment by junghans** [comment]
Most of the current tests were just examples, which I turned into regression tests, half of them compare against some reference data, but by no means are these unittests!

**Comment by acfogarty** [comment]
When I first started writing unit tests for recent commits, my main confusion was that half the stuff in the testsuite didn't seem to actually BE unit tests... Now I get it!
I think @MrTheodor recent commits are good examples of unittests for us all to copy. What do you all think?

**Comment by junghans** [comment]
Agreed! Long term, I would like to separate unit and regression tests, but for now mixed tests are better than no tests.

**Comment by niktre** [comment]
1. How short should the tests be? It there an upper limit? 1 sec?
2. Should they test "physics" or not? An example: at the moment I have an LB test that runs small stochastic LB liquid and calculates after short run the average density and momentum. Then I compare them with the initial ones, so that I know, that there is no mass or momentum leak in the system. Is this test ok?

**Comment by MrTheodor** [comment]
@niktre Currently travis has a limit of 50min for build; but I wouldn't restrict time at this stage. We don't have problems with long tests but with missing tests. 
Ad 2. I think that's important that the physics behind is correctly tested, this should be also the case of the interactions. As we deal with mixed environment and the tests run on Python level, we actually test the interface and the results of operations at that level.
